### PR TITLE
Add option to format time in mesh layer

### PR DESCRIPTION
**Pull request by PeterPetrik**
https://issues.qgis.org/issues/20731

### CODE REVIEW

**Review by wonder-sk — APPROVED** [review]
The code looks good, just needs some indentation/spelling fixes.

It would be nice though to add few tests to check formatted relative/absolute time output...

**Comment by rduivenvoorde** [comment]
@PeterPetrik cool feature thanks! FYI I'm fiddling with the creation of netcdf via python. According to eg: https://www.unidata.ucar.edu/software/netcdf/time/recs.html they always have some 'time.units" attribute and then data, so eg I I created (see attached tiny nc file):

[test4dz.zip](https://github.com/qgis/QGIS/files/2949888/test4dz.zip)

```
times.units = "hours since 0001-01-01 00:00:00.0"
times_data = [17533056., 17533068., 17533080.]
```

Then in QGIS (after selecting absolute time, I see: 

gnome-shell-screenshot-6XJKYZ

So 3 times the same date. 
Also note (bug or is it because the data is wrong interpreted) that the date is not changing if you move the timeslider.

While in panoply you see:

gnome-shell-screenshot-BJQLYZ

Or is the (very free) interpreting of this time.units from netcdf not yet implemented?

**Comment by rduivenvoorde** [comment]
I'll do some more tests at home. If interested, this is a small zip file to create these nc files, and also containing some links to nc example code:

[create_nc.zip](https://github.com/qgis/QGIS/files/2949899/create_nc.zip)

But plz let me know if this is not yet supposed to work.

File: python/core/auto_generated/mesh/qgsmeshtimesettings.sip.in [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+Represents a mesh time settings for mesh datasets
+
+.. versionadded:: 3.8
+%End
+
+%TypeHeaderCode
+#include "qgsmeshtimesettings.h"
+%End
+  public:
+    QDomElement writeXml( QDomDocument &doc ) const;
```
Comment (nyalldawson) [inline python/core/auto_generated/mesh/qgsmeshtimesettings.sip.in]: Should these methods also be taking a QgsReadWriteContext for possible future use?

**Comment by nyalldawson** [comment]
Is there anything particularly mesh related in the core stuff? I wonder if this class has potential usage elsewhere

**Comment by roya0045** [comment]
Could this be extended to other types of data. I know that some people use netcdf to store temporal data that is associated with vector features and I was curious to know if the core could be use to change the symbology/symbology field of a given feature vector (linked to a netcdf or not) in order to animate things such as traffic or water. (Maybe this is already doable and I never found it.)

**Comment by PeterPetrik** [comment]
@nyalldawson 
>> Is there anything particularly mesh related in the core stuff? I wonder if this class has potential usage elsewhere
Probably yes, when we implement that time support for other types. It would be easy to move when necessary.

**Comment by PeterPetrik** [comment]
@roya0045 definitely, but that would be a huge effort outside scope of this PR, see https://github.com/qgis/QGIS-Enhancement-Proposals/issues/128

**Comment by PeterPetrik** [comment]
@rduivenvoorde can you please try with "hours since 2010-01-01 00:00:00.0" or something that 5th significant digit in "hours" representation is valid (look at time in Metadata , all the times are parsed to the same double value 1.7533e7). This is probably limitation of the MDAL and time parsing created https://github.com/lutraconsulting/MDAL/issues/98

**Comment by PeterPetrik** [comment]
@pcav @rduivenvoorde for automatic set of reference time from source dataset we need to implement https://github.com/lutraconsulting/MDAL/issues/100

**Comment by rduivenvoorde** [comment]
Ah ok, now I understand your question :-) The fact that I used such strange huge numbers (hours since 0) made that we did not pick it up as different.

Create nc file now with much smalller time-data numbers. Relative time looks fine now:

gnome-shell-screenshot-4SATYZ

**Comment by saberraz** [comment]
Can we have OK, Cancel buttons?
I am trying to manually change the reference date instead of using the calendar widget and it does not take my changes.

**Comment by PeterPetrik** [comment]
> Can we have OK, Cancel buttons?
> I am trying to manually change the reference date instead of using the calendar widget and there is no way of applying my changes.

can you please create new issue on this? Calendar is Qt widget, so not sure if I can change that. But maybe there are some options on it too.

**Pull request merged**